core/server/worker/src/main/java/alluxio/worker/block/UnderFileSystemBlockStore.java

```diff
@@ -0,0 +1,391 @@
+ * either express or implied, as more fully set forth in the License.
+ *
+ * See the NOTICE file distributed with this work for information regarding copyright ownership.
+ */
+
+package alluxio.worker.block;
+
+import alluxio.exception.BlockAlreadyExistsException;
+import alluxio.exception.BlockDoesNotExistException;
+import alluxio.exception.ExceptionMessage;
+import alluxio.worker.block.io.BlockReader;
+import alluxio.worker.block.io.BlockWriter;
+import alluxio.worker.block.meta.UnderFileSystemBlockMeta;
+import alluxio.worker.block.options.OpenUfsBlockOptions;
+
+import com.google.common.base.Objects;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.io.IOException;
+import java.util.HashMap;
+import java.util.HashSet;
+import java.util.Map;
+import java.util.Set;
+import java.util.concurrent.locks.ReentrantLock;
+
+import javax.annotation.concurrent.GuardedBy;
+
+/**
+ * This class manages the virtual blocks in the UFS for delegated UFS reads/writes.
+ *
+ * The usage pattern:
+ *  acquireAccess(blockMeta, maxConcurrency)
+ *  cleanup(sessionId, blockId)
+ *  releaseAccess(sessionId, blockId)
+ *
+ * If the client is lost before releasing or cleaning up the session, the session cleaner will
+ * clean the data.
+ */
+public final class UnderFileSystemBlockStore {
+  private static final Logger LOG = LoggerFactory.getLogger(UnderFileSystemBlockStore.class);
+
+  /**
+   * This lock protects mBlocks, mSessionIdToBlockIds and mBlockIdToSessionIds. For any read/write
+   * operations to these maps, the lock needs to be acquired. But once you get the block
+   * information from the map (e.g. mBlocks), the lock does not need to be acquired. For example,
+   * the block reader/writer within the BlockInfo can be updated without acquiring this lock.
+   * This is based on the assumption that one session won't open multiple readers/writers on the
+   * same block. If the client do that, the client can see failures but the worker won't crash.
+   */
+  private final ReentrantLock mLock = new ReentrantLock();
+  @GuardedBy("mLock")
+  /** Maps from the {@link Key} to the {@link BlockInfo}. */
+  private final Map<Key, BlockInfo> mBlocks = new HashMap<>();
+  @GuardedBy("mLock")
+  /** Maps from the session ID to the block IDs. */
+  private final Map<Long, Set<Long>> mSessionIdToBlockIds = new HashMap<>();
+  @GuardedBy("mLock")
+  /** Maps from the block ID to the session IDs. */
+  private final Map<Long, Set<Long>> mBlockIdToSessionIds = new HashMap<>();
+
+  /** The Local block store. */
+  private final BlockStore mLocalBlockStore;
+
+  /**
+   * Creates an instance of {@link UnderFileSystemBlockStore}.
+   *
+   * @param localBlockStore the local block store
+   */
+  public UnderFileSystemBlockStore(BlockStore localBlockStore) {
+    mLocalBlockStore = localBlockStore;
+  }
+
+  /**
+   * Acquires access for a UFS block given a {@link UnderFileSystemBlockMeta} and the limit on
```

Comment:
can you describe this method with more details? e.g., what is the max concurrency? (I think you should explain it here that we allow certain number of sessions to read this ufs block by themselves; after this number, they will not be granted and etc.